README.md
## Gitlab Cookbook

[![Build Status](https://travis-ci.org/atomic-penguin/cookbook-gitlab.png?branch=master)](https://travis-ci.org/atomic-penguin/cookbook-gitlab)

This cookbook will deploy gitlab; a free project and repository management
application.

Gitlab code hosted on github [here](https://github.com/gitlabhq/gitlabhq).

## Important changes

Going forward, the cookbook major version (i.e. 6.1.x) will target the
matching stable branch (i.e. 6-1-stable) of the Gitlab application.
The 6.1.x release is not backwards compatible with previous versions targeting
Gitlab master.

For the initial 6.1.x release of the cookbook, the default Ruby will
be 1.9.3 compiled with [ruby_build](http://fnichol.github.com/chef-ruby_build/).
Using a compiled 1.9.3 Ruby follows the Gitlab installation guidelines upstream.
If you have a better approach which reduced complexity or reduces converge time, 
please open a pull request on Github.

The application home has moved from `/var/gitlab/gitlab` to `/srv/git/gitlab`
in accordance with the [Filesystem Hierarchy Standard (FHS) version 2.3](http://www.pathname.com/fhs/).

## Requirements
============

* Hard disk space
  - About 600 Mb, plus enough space for repositories in application home

* You need to add `mysql::server` or `postgresql::server` to your run\_list
  if you intend to run the whole application stack on a single instance.
  The `gitlab::mysql` and `gitlab::postgresql` recipes set up the RDBMS
  connection only, to allow for multi-instance architecture.  See
  `Database Attributes` documentation below for RDBMS connection
  parameters.

## Cookbook dependencies
============

* [ruby\_build](http://fnichol.github.com/chef-ruby_build/)
  - Thanks to Fletcher Nichol for his awesome ruby\_build cookbook.
    This ruby\_build LWRP is used to build Ruby 1.9.3 for gitlab.

* [redisio](http://community.opscode.com/cookbooks/redisio)
  - Thanks to Brian Bianco for this Redis cookbook.

* Opscode, Inc cookbooks
  - [git](http://community.opscode.com/cookbooks/git)
  - [build-essential](http://community.opscode.com/cookbooks/build-essential)
  - [sudo](http://community.opscode.com/cookbooks/sudo)
  - [nginx](http://community.opscode.com/cookbooks/nginx)
  - [logrotate](http://community.opscode.com/cookbooks/logrotate)
  - [openssh](http://community.opscode.com/cookbooks/openssh)
  - [perl](http://community.opscode.com/cookbooks/perl)
  - [xml](http://community.opscode.com/cookbooks/xml)
  - [zlib](http://community.opscode.com/cookbooks/zlib)
  - [database](http://community.opscode.com/cookbooks/database)


Attributes
==========

* `gitlab['user']` & `gitlab['group']`
  - Gitlab service user and group for Unicorn Rails app, default `git`

* `gitlab['home']`
  - Gitlab top-level home for service account, default `/srv/git`

* `gitlab['app_home']`
  - Gitlab application home, default `/srv/git/gitlab`

* `gitlab['email_from']`
  - Gitlab email from, default `gitlab@ + node.fqdn`

* `gitlab['support_email']`
  - Gitlab support email, default `gitlab-support@ + node.fqdn`

* `gitlab['git_url']`
  - Github gitlab address, default git://github.com/gitlabhq/gitlabhq.git

* `gitlab['git_branch']`
  - Defaults to stable GitlabHQ branch matching the major version of this cookbook. e.g. 6.1.x => 6-1-stable

* `gitlab['packages']`
  - Platform specific OS packages

* `gitlab['trust_local_sshkeys']`
  - `ssh_config` key for gitlab to trust localhost keys automatically, default yes

* `gitlab['install_ruby']`
  - Attribute to determine whether vendor packages are installed,
    or Rubies are built, defaults 1.9.3 for Debian and RHEL family platforms.
  - If you choose to use a vendor provided package, you will need to use
    a role to override the `gitlab['packages']` array.

* `gitlab['https']`
  - Whether https should be used. Default false

* `gitlab['self_signed_cert']`
  - Allows self-signed certificates over https protocol. Default false

* `gitlab['certificate_databag_id']`
  - Encrypted databag name containing certificate file, CA bundle, and key. Default nil
  - See [certificate cookbook](http://community.opscode.com/cookbooks/certificate) for further information.

* `gitlab['backup_path']`
  - Path in file system where backups are stored. Default `gitlab['app_home'] + backups/`

* `gitlab['backup_keep_time']`
  - Units are seconds. Older backups will automatically be deleted when new backup is created. Set to 0 to keep backups forever.
  - Defaults to 604800

* `gitlab['listen_ip']`
  - IP address that nginx will listen on, default `*` (listen on all IPs)

* `gitlab['listen_port']`
  - Port that nginx will listen on, default to 80 if gitlab['https'] is set to false, 443 if set to true

* `gitlab['web_fqdn']`
  - An overridable service name, used in gitlab and unicorn configuration files.
    Useful if `hostname -f` is not the same as the customer facing hostname.
    Default is unset. Effective default is node['fqdn']

* `gitlab['nginx_server_names']`
  - An array with nginx `server_name` matches.  Helpful to override default test site pages
    shipping with some nginx packages.  Default `[ 'gitlab.*', node['fqdn'] ]`.
    See [nginx server_name documentation](http://nginx.org/en/docs/http/server_names.html)
    for valid matching patterns.

* gitlab['gravatar']['enabled']
  - Use Gravatar to fetch user avatars 
  - Options: "true", "false"
  - Default "true"

### User privileges 

**Note**, This attributes are useful when you want only admins to create projects and groups. And to restrict username changing.

* default['gitlab']['default_projects_limit']
  - When you create a user it takes this value as his **projects_limit** profile setting. 
  - If you put it at 0, regular users can't create repos.

* default['gitlab']['default_can_create_group'] = true
  - When you create a user it takes this value as his **can_create_group** profile setting.
  - Name is pretty explanatory about what it means :)

* default['gitlab']['username_changing_enabled'] = false
  - Disable users changing his username.
  - This is **true** on a default gitlab install. 

### Database Attributes

**Note**, most of the database attributes have sane defaults. You will only need to change these configuration options if
you're using a non-standard installation. Please see `attributes/default.rb` for more information on how a dynamic attribute
is calculated.

* gitlab['database']['type']
  - The database (datastore) to use.
  - Options: "mysql", "postgres"
  - Default "mysql"

* gitlab['database']['adapter']
  - The Rails adapter to use with the database type
  - Options: "mysql2", "postgresql"
  - Default (varies based on `type`)

* gitlab['database']['encoding']
  - The database encoding
  - Default (varies based on `type`)

* gitlab['database']['host']
  - The host (fqdn) where the database exists
  - Default `localhost`

* gitlab['database']['userhost']
  - The host (fqdn) from which the database user may connect.
  - Default `localhost`

* gitlab['database']['pool']
  - The maximum number of connections to allow
  - Default 5

* gitlab['database']['database']
  - The name of the database
  - Default `gitlab`

* gitlab['database']['username']
  - The username for the database
  - Default `gitlab`

### LDAP Authentication Attributes

* `gitlab'['ldap']['enabled']`
  - Use LDAP for authentication
  - Default: false

* `gitlab['ldap']['host']`
  - Hostname of your LDAP server
  - Default: "_your_ldap_server"

* `gitlab['ldap']['base']`
  - Base DN for users (e.g. dc=users,dc=example,dc=com)
  - Default: "_the_base_where_you_search_for_users"

* `gitlab['ldap']['port']`
  - LDAP server port
  - Default: 636

* `gitlab['ldap']['uid']`
  - User ID used when searching for users (e.g. uid, cn, or sAMAccountName)
  - Default: "sAMAccountName"

* `gitlab['ldap']['method']`
  - Connection method used with LDAP server
  - Options: "ssl", "plain"
  - Default: "ssl"

* `gitlab['ldap']['bind_dn']`
  - Some servers require a username in order to bind.
  - Leave default if your server supports anonymous bind.
  - Default: "_the_full_dn_of_the_user_you_will_bind_with"

* `gitlab['ldap']['password']`
  - Some servers require a password in order to bind
  - Leave default if your server supports anonymous bind.
  - Default: "_the_password_of_the_bind_user"

* `gitlab['ldap']['allow_username_or_email_login']`
  - If you want to allow users to login using both username and email address even though username (uid) will actually be used
  - Default: true

Usage
=====

Optionally override application paths using gitlab['git\_home'] and gitlab['home'].

Add recipe gitlab::default to run\_list.  Go grab a lunch, or two, if Ruby has to build.

The default admin credentials for the gitlab application are as follows:

    User: admin@local.host
    Password: 5iveL!fe

Of course you should change these first thing, once deployed.

## Role example for Gitlab with https, and MySQL

```
name "gitlab_https"
description "Configures and installs gitlab w/ https, and mysql server"
override_attributes "gitlab" => {
  "https" => true,
  "certificate_databag_id" => "wildcard"
}
run_list "recipe[mysql::server]", "recipe[gitlab]
```

License and Author
==================

Author: Gerald L. Hevener Jr., M.S.
Copyright: 2012

Author: Eric G. Wolfe
Copyright: 2012

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
Gitlab Cookbook Changelog
=========================

v6.4.3
------

* Include new user privileges attributes.
* Some README.md points on LDAP auth when anonymous binding is enabled on server.
* Fix Path issue when executing bundle. It was using chef-client embedded ruby instead gitlab one.
  - Just adding the path to the giltab bin directory at the beginning of the command.

v6.4.2
------

* Add a profile script shim, so init script can correct Ruby.

v6.4.1
------

* Issue #60 - thanks to @nickryand
  - Modified the bundle install command to drop a file on successful
    completion 
  - Added the absolute path to the bundler binary installed into the
    system ruby path.
  - Added a more accurate pattern matcher to gitlab service resource
    so Chef can find unicorn_rails processes instead of finding processes that
    have the 'gitlab' string in them.
  - Changed the background call to script/web and script/background_jobs
    in the startup script.  These are now foreground calls forcing the start
    script to block until they return (after the processes are up).  This
    prevents the Chef run from completing before the unicorn_rails processes
    are up.

* Other fixes/tweaks
  - Remove yum dependency hell.
  - Change init/database.yml order due to service subscription.
  - Add a :80 redirect to :443, avoids default site showing on :80
    when `gitlab['https']` set.

* Update test files
  - Pin omnibus version at 11.8.2, redisio remote file resource failure.
  - Remove explicit nil on databag_certificate_id, fails on default suite 
  - Remove support Gemfile, and update travis.yml
  - Add bats tests for default/https suites

v6.4.0
------

  * Bump gitlab-shell to v1.8.0
  * Bump gitlab to 6.4.0
  * Remove stale nginx config file
  * Default recipe changes
    - Update gitlab-shell config to point at /usr/local/bin/redis-cli
    - Fix permissions on .ssh/authorized_keys
    - Add rack_attack file
    - Add precompile assets execute
    - Add logrotate dependency and configuration 

v6.1.21
-------

  * Add node['database']['userhost'] attribute, fixes #57
  * Fix gitlab['listen_port']/gitlab['https'] condition, fixes #58
  * Add node['gitlab']['self_signed_cert'] and documentation, fixes #58
  * Revert yum-epel changes, COOK-4164 blocking, will re-open
    - Fix dependency hell w/ nginx cookbook
  * Add rubocop linting
  * Add chefspec mysql/postgres http/https branching specs
  * Remove unneeded python dependency 

v6.1.10
-------

Bugs squashed:

  * Re-order database components to fix Ubuntu Test-Kitchen run. #44
  * Update port forwarding configuration in kitchen.yml. #46
  * Update Satellite paths to fix Gitlab forking. #50
  * ruby-build installs to node['gitlab']['home'], avoiding PATH hacks
    for Rake.  Use node['gitlab']['install_ruby_path'] to override. #16
  * ruby-build failing on 1.9.3-p448, bump version to 1.9.3-p484.
  * Pinned yum dependency on < 3.0.0

Features added:

  * Add knob for disabling Gravatar. #51
  * Add LDAP Authentication support. #49

Removed:

  * sudo cookbook dependency removed. #52

v6.1.0
------

Clean up some of the code to target [6-1-stable branch](https://github.com/gitlabhq/gitlabhq/blob/6-1-stable/doc/install/installation.md)

* Move gitlab.home to /srv/git - [FHS 2.3](http://www.pathname.com/fhs/pub/fhs-2.3.html)
* Use ruby_build to compile 1.9.3 by default per gitlabhq documentation.
* Clean up both cookbook and package dependencies.
* Remove ruby-shadow, included in Omnibus, not needed anyway as no
  password is set.
* Use gitconfig template, instead of execute.
* Add test-kitchen skeleton, and certificate data_bag integration.
  - Add gitlab.certificate_databag_id to deploy certificate from encrypted databag.
* Minor sudo fix in gitlab init script.
* Use nginx_site definition to disable default nginx site.
* Add nginx_server_names array for hostname match precedence over potential default sites matching `_`.

v0.5.0
------

* Database back-end changes via @sethvargo
  - Adds a mysql and postgresql database creation 
  - **Breaking change**: developed against githlabhq/master (sqlite no longer supported)

* Gitolite deploy changes via @dscheu
  - Deploy gitlabhq/gitolite, not sitaramc/gitolite

* Add configurable backup paths to Gitlab configuration via @dscheu

v0.3.4
------

Fix issues with stable snapshot v3.x
    
* Avoid installing pg, gem which adds extra dependencies
* Add change to default gitolite.rc per upgrade instructions

v0.3.3
------

Issues #9 and #10

Issue 9: this version MAY fix issues with key generation when
cookbook is invocated via chef-solo.  There may exist other
chef-solo blockers within the cookbook.  Specifically, `File.exists?`
guards were added to SSH public key generation code blocks.

Issue 10: this version fixes one minor dependency bug with EPEL
dependencies via metadata and inclusion.

v0.3.2
------

* Default gitlab branch to stable

v0.3.1
------

* ISSUE 7: public key template fails to render
* ISSUE 8: unicorn_rails script fails on ruby package platforms 

v0.3.0
------

* Missing bracket
* Change single-quote variable to symbol notation
* install python and symlink redis-cli so hooks work
* HTTPS options for nginx
* Ubuntu/Debian platform compatibility fixes
* [FC035](http://acrmp.github.com/foodcritic/#FC035): Template uses node attribute directly

v0.2.1
------

  Thanks to Fletcher Nichol for the feedback and fixes :)

  * Add `gitlab_branch` attribute.
  * Fix directory block syntax (do).

v0.2.0
------

  * Epic public release <crowd cheers>
  * Moar testing
  * Clean up init script
  * Fix unicorn config
  * Fix gitlab home permissions for nginx

v0.1.0
------

  #Epic refactor

  * Write long README
  * variable renaming to simplify readability
  * refactor dependencies and package lists
  * generate ssh keys in Ruby, import to gitolite
  * Integrate gitolite recipe into cookbook 
  * Fix broken ssh problems
  * fixup git home permissions

recipes/default.rb
#
# Cookbook Name:: gitlab
# Recipe:: default
#
# Copyright 2012, Gerald L. Hevener Jr., M.S.
# Copyright 2012, Eric G. Wolfe
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

case node['platform_family']
when 'debian'
  include_recipe 'apt'
when 'rhel'
  include_recipe 'yum-epel'
end

# Setup the database connection
case node['gitlab']['database']['type']
when 'mysql'
  include_recipe 'gitlab::mysql'
when 'postgres'
  include_recipe 'gitlab::postgres'
else
  Chef::Log.error "#{node['gitlab']['database']['type']} is not a valid type. Please use 'mysql' or 'postgres'!"
end

# Install the required packages via cookbook
node['gitlab']['cookbook_dependencies'].each do |requirement|
  include_recipe requirement
end

# Install required packages for Gitlab
node['gitlab']['packages'].each do |pkg|
  package pkg
end

# Add a git user for Gitlab
user node['gitlab']['user'] do
  comment 'Gitlab User'
  home node['gitlab']['home']
  shell '/bin/bash'
  supports manage_home: true
end

# Fix home permissions for nginx
directory node['gitlab']['home'] do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0755'
end

# Create a $HOME/.ssh folder
directory "#{node['gitlab']['home']}/.ssh" do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0700'
end

file "#{node['gitlab']['home']}/.ssh/authorized_keys" do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0600'
end

# Drop off git config
template "#{node['gitlab']['home']}/.gitconfig" do
  source 'gitconfig.erb'
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
end

# Configure gitlab user to auto-accept localhost SSH keys
template "#{node['gitlab']['home']}/.ssh/config" do
  source 'ssh_config.erb'
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
  variables(
      fqdn: node['fqdn'],
      trust_local_sshkeys: node['gitlab']['trust_local_sshkeys']
  )
end

# The recommended Ruby is >= 1.9.3
# We'll use Fletcher Nichol's ruby_build cookbook to compile a Ruby.
if node['gitlab']['install_ruby'] !~ /package/
  ruby_build_ruby node['gitlab']['install_ruby'] do
    prefix_path node['gitlab']['install_ruby_path']
    user node['gitlab']['user']
    group node['gitlab']['user']
  end

  # Install required Ruby Gems for Gitlab with ~git/bin/gem
  %w[charlock_holmes bundler].each do |gempkg|
    gem_package gempkg do
      gem_binary "#{node['gitlab']['install_ruby_path']}/bin/gem"
      action :install
      options('--no-ri --no-rdoc')
    end
  end
else
# Install required Ruby Gems for Gitlab with system gem
  %w[charlock_holmes bundler].each do |gempkg|
    gem_package gempkg do
      gem_binary "#{node['gitlab']['install_ruby_path']}/bin/gem"
      action :install
      options('--no-ri --no-rdoc')
    end
  end
end

# setup gitlab-shell
# Clone Gitlab-shell repo
git node['gitlab']['shell']['home'] do
  repository node['gitlab']['shell']['git_url']
  reference node['gitlab']['shell']['git_branch']
  action :checkout
  user node['gitlab']['user']
  group node['gitlab']['group']
end

# Either listen_port has been configured elsewhere or we calculate it depending on the https flag
listen_port = node['gitlab']['listen_port'] || (node['gitlab']['https'] ? 443 : 80)

# Address of gitlab api for which gitlab-shell should connect
api_fqdn = node['gitlab']['web_fqdn'] || node['fqdn']

# render gitlab-shell config
template node['gitlab']['shell']['home'] + '/config.yml' do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
  source 'shell_config.yml.erb'
  variables(
      fqdn: api_fqdn,
      listen: listen_port
  )
end

# Clone Gitlab repo from github
git node['gitlab']['app_home'] do
  repository node['gitlab']['git_url']
  reference node['gitlab']['git_branch']
  action :checkout
  user node['gitlab']['user']
  group node['gitlab']['group']
end

# Drop off a profile script to be included in init script
template "#{node['gitlab']['home']}/.profile" do
  source 'profile_gitlab.sh.erb'
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0755'
end

# Render gitlab init script
# This needs to happen before gitlab.yml is rendered.
# So when the service is subscribed, the init file will be in place
template '/etc/init.d/gitlab' do
  owner 'root'
  group 'root'
  mode '0755'
  source 'gitlab.init.erb'
  variables(
      gitlab_app_home: node['gitlab']['app_home'],
      gitlab_user: node['gitlab']['user']
  )
end

# Write the database.yml
template "#{node['gitlab']['app_home']}/config/database.yml" do
  source 'database.yml.erb'
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
  variables(
      adapter: node['gitlab']['database']['adapter'],
      encoding: node['gitlab']['database']['encoding'],
      host: node['gitlab']['database']['host'],
      database: node['gitlab']['database']['database'],
      pool: node['gitlab']['database']['pool'],
      username: node['gitlab']['database']['username'],
      password: node['gitlab']['database']['password']
  )
end

# Render gitlab config file
template "#{node['gitlab']['app_home']}/config/gitlab.yml" do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
  variables(
      fqdn: node['gitlab']['web_fqdn'] || node['fqdn'],
      https_boolean: node['gitlab']['https'],
      git_user: node['gitlab']['user'],
      git_home: node['gitlab']['home'],
      backup_path: node['gitlab']['backup_path'],
      backup_keep_time: node['gitlab']['backup_keep_time']
  )
end

# Copy file rack_attack.rb
cookbook_file "#{node['gitlab']['app_home']}/config/initializers/rack_attack.rb" do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
end

# create log, tmp, pids and sockets directory
%w{ log tmp tmp/pids tmp/sockets public/uploads }.each do |dir|
  directory File.join(node['gitlab']['app_home'], dir) do
    user node['gitlab']['user']
    group node['gitlab']['group']
    mode '0755'
    recursive true
    action :create
  end
end

# logrotate gitlab-shell and gitlab
logrotate_app 'gitlab' do
  frequency 'weekly'
  path ["#{node['gitlab']['app_home']}/log/*.log",
        "#{node['gitlab']['shell']['home']}/gitlab-shell.log"]
  rotate 52
  options %w[compress delaycompress notifempty copytruncate]
end

# create gitlab-satellites directory
directory File.join(node['gitlab']['home'], 'gitlab-satellites') do
  user node['gitlab']['user']
  group node['gitlab']['group']
  mode '0755'
  recursive true
  action :create
end

# create repositories directory
directory File.join(node['gitlab']['home'], 'repositories') do
  user node['gitlab']['user']
  group node['gitlab']['group']
  mode '2770'
  recursive true
  action :create
end

# create backup_path
directory node['gitlab']['backup_path'] do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '00755'
  action :create
end

# Render unicorn template
template "#{node['gitlab']['app_home']}/config/unicorn.rb" do
  owner node['gitlab']['user']
  group node['gitlab']['group']
  mode '0644'
  variables(
      fqdn: node['fqdn'],
      gitlab_app_home: node['gitlab']['app_home']
  )
end

without_group = node['gitlab']['database']['type'] == 'mysql' ? 'postgres' : 'mysql'

bundler_binary = "#{node['gitlab']['install_ruby_path']}/bin/bundle"
bundle_success = "#{node['gitlab']['app_home']}/vendor/bundle/.success"

# Install Gems with bundle install
execute 'gitlab-bundle-install' do
  command "#{bundler_binary} install --deployment --without development test #{without_group} aws && touch #{bundle_success}"
  cwd node['gitlab']['app_home']
  user node['gitlab']['user']
  group node['gitlab']['group']
  environment('LANG' => 'en_US.UTF-8', 'LC_ALL' => 'en_US.UTF-8')
  not_if { File.exists?(bundle_success) }
end

# Precompile assets
execute 'gitlab-bundle-precompile-assets' do
  command "#{bundler_binary} exec rake assets:precompile RAILS_ENV=production"
  cwd node['gitlab']['app_home']
  user node['gitlab']['user']
  group node['gitlab']['group']
  environment('LANG' => 'en_US.UTF-8', 'LC_ALL' => 'en_US.UTF-8')
  only_if { Dir["#{node['gitlab']['app_home']}/public/assets/*"].empty? }
end

# Initialize database
execute 'gitlab-bundle-rake' do
  command "#{bundler_binary} exec rake gitlab:setup RAILS_ENV=production force=yes && touch .gitlab-setup"
  cwd node['gitlab']['app_home']
  user node['gitlab']['user']
  group node['gitlab']['group']
  not_if { File.exists?("#{node['gitlab']['app_home']}/.gitlab-setup") }
end

# Use certificate cookbook for keys
certificate_manage "#{node['gitlab']['certificate_databag_id']}" do
  cert_path '/etc/nginx/ssl'
  owner node['gitlab']['user']
  group node['gitlab']['user']
  nginx_cert true
  only_if { node['gitlab']['https'] && !node['gitlab']['certificate_databag_id'].nil? }
end

# Create nginx directories before dropping off templates
include_recipe 'nginx::commons_dir'

# Render and activate nginx default vhost config
template '/etc/nginx/sites-available/gitlab' do
  owner 'root'
  group 'root'
  mode '0644'
  source 'nginx.gitlab.erb'
  notifies :restart, 'service[nginx]'
  variables(
      server_name: node['gitlab']['nginx_server_names'].join(' '),
      hostname: node['hostname'],
      gitlab_app_home: node['gitlab']['app_home'],
      https_boolean: node['gitlab']['https'],
      ssl_certificate: node['gitlab']['ssl_certificate'],
      ssl_certificate_key: node['gitlab']['ssl_certificate_key'],
      listen: "#{node['gitlab']['listen_ip']}:#{listen_port}"
  )
end

# Install nginx
include_recipe 'nginx'

# Enable gitlab site
nginx_site 'gitlab' do
  enable true
end

# Disable default site
nginx_site 'default' do
  enable false
end

# Enable and start unicorn and sidekiq service
service 'gitlab' do
  priority 30
  pattern "unicorn_rails master -D -c #{node['gitlab']['app_home']}/config/unicorn.rb"
  action [:enable, :start]
  subscribes :restart, "template[#{node['gitlab']['app_home']}/config/gitlab.yml]", :delayed
end
